w3-blob.md

```diff
@@ -0,0 +1,419 @@
+
+1. Provided **sub**ject space is not provisioned with a provider.
+1. Provided `blob.size` is outside of supported range.
+1. Provided `blob.content` is not a valid [multihash].
+1. Provided `blob.content` [multihash] hashing algorithm is not supported.
+
+Invocation MUST succeed if non of the above is true. Success value MUST be an object with a `claim` field set to [await/ok] of the task that produces [location claim].
+
+Task linked from the `claim` of the success value MUST be present in the receipt effects _(`next` field)_.
+
+#### Add Blob Effects
+
+Successful invocation MUST start a workflow consisting of following tasks, that MUST be set in receipt effects (`next` field) in the following order.
+
+1. [Allocate Blob]
+1. [Accept Blob]
+
+## Allocate Blob
+
+Authorized agent MAY invoke `/service/blob/allocate` capability on the [provider] subject to create a memory address where `blob` content can be written via HTTP `PUT` request.
+
+### Allocate Blob Capability
+
+#### Allocate Blob Capability Schema
+
+```ts
+type BlobAllocate = {
+  cmd:  "/service/blob/allocate"
+  sub:  ProviderDID
+  args: {
+    space:  SpaceDID
+    blob:   Blob
+    cause:  Link<AddBlob>
+  }
+}
+```
+
+#### Allocation Space
+
+The `args.space` field MUST be set to the [DID] of the user space where allocation takes place.
+
+#### Allocation Blob
+
+The `args.blob` field MUST be set to the `Blob` the space is allocated for.
+
+#### Allocation Cause
+
+The `args.cause` field MUST be set to the [Link] for an [Add Blob] task, that caused an allocation.
+
+### Allocate Blob Receipt
+
+Allocations MUST fail if `space` does not have enough capacity for the `blob` and succeed otherwise.
+
+#### Allocate Blob Receipt Schema
+
+```ts
+type BlobAllocateReceipt = {
+  ran:  Link<BlobAllocate>
+  out:  Result<BlobAllocateOk, BlobAllocateError>
+  next: []
+}
+
+type BlobAllocateOk = {
+  size: int
+  address?: BlobAddress
+}
+
+type BlobAddress = {
+  url:     string
+  headers: {[key:string]: string}
+}
+```
+
+### Allocation Size
+
+The `out.ok.size` MUST be set to the number of bytes that were allocated for the `Blob`. It MUST be equal to either:
+
+1. The `args.blob.size` of the invocation.
+2. `0` if space already has memory allocated for the `args.blob`.
+
+### Allocation Address
+
+The optional `out.ok.address` SHOULD be omitted when content for the allocated is already available on site. Otherwise it MUST be set to the `BlobAddress` that can receive a blob content.
+
+The `url` of the `BlobAddress` MUST be an HTTP(S) location that can receive blob content via HTTP `PUT` request, as long as HTTP headers from `headers` dictionary are set on the request.
+
+It is RECOMMENDED that issued `BlobAddress` only accept `PUT` payload that matches requested `blob`, both content [multihash] and size.
+
+> ℹ️ If enforcing above recommendation is not possible implementation MUST enforce in [Accept Blob] invocation.
+
+### Allocation Effects
+
+Allocation MUST have no effects.
```

Comment:
I was thinking this would actually point to Accept as well, given next step is that after allocation. Why not?

Reply:
Perhaps 🤔, but here here is my sentiment

Can I allocate space without waiting for it to be received ? If so, this is a function to do and it should not have a continuation there. If no, I think I need bit more convincing because here are my counter arguments

1. Overall I think you could roughly translate blob/add as following JS code

   ```js
   const blobAdd = async (task) => {
      const { sub, args: { blob, exp } } = task
      // check that we have billing setup for the space
      const { ok: subscription, error: cantbill } = await service.getSubscription(sub)
      if (cantbill) {
         return { error: cantbill }
      }

      // allocate memory for the blob
      const { ok: memory, error: allocError } = await service.allocate(sub, { exp, blob, cause: task.link() })
      if (allocError) {
        return { error: allocError }
      }
      
      // add usage to the billing
      if (memory.size > 0) {
         await subscription.bill.add({ storage: size, cause: task.link() })
      }
      
       // if we have memory address we need to await for the content
      if (memory.address) {
         return { ok: await accept({ blob, exp  }) }
      } else {
         return { ok: findReceiptFor({ accept, blob, exp }) }
      }
    }

    const accept = async ({ blob, exp }) => {
       const { ok: content, error: timeout } = await waitForPUT({ blob })
        if (timeout) {
           return { error: timeout }
        }
          
        // validate content matches the blob
        const { error } = await vaildate(content, blob)
        if (error) {
            return { error }
        }

        return { ok: createLocationClaim(blob) })
    }
    ``` 

    And through above lens, `addBlob` is a task that coordinates subtasks, but subtasks `allocate` and `accept` are independent of each other.

2. If we made `accept` effect of the `allocate`, then we probably should not make it effect of the `blobAdd` also. However making it effect of the `blobAdd` is better because we can then draw the whole execution graph from the receipt without having to follow each effect.


For what it's worth these observations and opinions are new to me also and  I have call with Fission on Monday to try and get their input on this as they spend a lot more time doing and implementing this.

Reply:
thanks for the details, I find this a good reasoning. My suggestion was more about having a receipt for a `blob/allocate` and not really linking it to anything. But we have as part of the task arguments the `blob/add` who caused this, so I think that also mitigates what I was looking for